### PR TITLE
Detect LDAP connection errors during `occ user:sync`

### PR DESCRIPTION
**Pull request by AlexAndBear**
## Description
Needs https://github.com/owncloud/user_ldap/pull/669 to work properly.

## Related Issue
- Fixes https://github.com/owncloud/enterprise/issues/4642

## How Has This Been Tested?

- test environment:
- test case 1:
- test case 2:
- ...

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Database schema changes (next release will require increase of minor version instead of patch)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests only (no source changes)

## Checklist:

- [x] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link> 
- [ ] Changelog item, see [TEMPLATE](https://github.com/owncloud/core/blob/master/changelog/TEMPLATE)

### CODE REVIEW

**Comment by jvillafanez** [comment]
I'm not sure how useful this solution will be. For the LDAP case, you have to search over all the results in order to count the users, which means that you're doing the same requests twice including pagination.

If we consider the network is unstable, counting could go wrong too. I'm not sure how it's handled, but unless the error is noticeable, the counting could say there are 500 users but you could sync 750 because there was an error while requesting the second page of the results for the count.

In addition, just an error message might not be enough. I've been told that the recommendation is to sync the users via (linux) cron job each 15-30 minutes, which means the command is expected to run automatically without user intervention, so it's unlikely that there will be anyone noticing that the command is showing an error message.

File: core/Command/User/SyncBackend.php [hunk 1]
```diff
@@ -228,24 +228,50 @@ private function syncMultipleUsers(
...
-			/* @phan-suppress-next-line PhanUndeclaredMethod */
-			$max = $backend->countUsers();
+
+		$toProcessUsersCount = null;
+
+		if ($backend->implementsActions(\OC_User_Backend::COUNT_USERS)) {
+			if ($input->getOption('seenOnly')) {
+				$toProcessUsersCount = 0;
+
+				$this->userManager->callForSeenUsers(function (IUser $user) use ($backend, &$toProcessUsersCount) {
```
Comment (JammingBen) [inline core/Command/User/SyncBackend.php:238]: @JanAckermann Maybe we can skip this ugly loop by calling `$toProcessUsersCount = $this->accountMapper->getUserCountPerBackend(true);`. I seems to return an array with all the available backends including their users.

**Comment by JammingBen** [comment]
> In addition, just an error message might not be enough. I've been told that the recommendation is to sync the users via (linux) cron job each 15-30 minutes, which means the command is expected to run automatically without user intervention, so it's unlikely that there will be anyone noticing that the command is showing an error message.

We already log the error, I think that's all we can do for now. I wouldn't go for active notifications like emails in case of a cron run, especially if the cron runs every 15 minutes.

**Comment by JanAckermann** [comment]
@jvillafanez 

>I'm not sure how useful this solution will be. For the LDAP case, you have to search over all the results in order to count the users, which means that you're doing the same requests twice including pagination.

Well, this is not good at all, I am not an expert in LDAP, I thought it would be like mysql `SELECT COUNT(*) FROM users`....
If it is not, well it breaks the whole solution.

>In addition, just an error message might not be enough. I've been told that the recommendation is to sync the users via (linux) cron job each 15-30 minutes, which means the command is expected to run automatically without user intervention, so it's unlikely that there will be anyone noticing that the command is showing an error message.

Errors are in logs, the issue is about running the command manually. 

So, I don't know how to solve it if we are not able to throw an Exception here

File: lib/private/User/Sync/BackendUsersIterator.php [hunk 2]
```diff
@@ -63,7 +72,13 @@ public function next() {
 		if ($this->hasMoreData && $this->dataPos >= $this->endPos) {
 			$this->page++;
 			$offset = $this->page * self::LIMIT;
+
 			$this->data = $this->backend->getUsers($this->search, self::LIMIT, $offset);
+
+			if($this->checkForConnectionErrors && method_exists($this->backend,'testConnection')){
```
Comment (JammingBen) [inline lib/private/User/Sync/BackendUsersIterator.php:78]:
```suggestion
			if($this->checkForConnectionErrors && method_exists($this->backend, 'checkForConnectionErrors')){
```

**Comment by JammingBen** [comment]
This is a "won't fix" for now: https://github.com/owncloud/enterprise/issues/4642#issuecomment-986586722

**Pull request closed without merging**